### PR TITLE
test_formatter: add {None:.2f} to print None, not {None}

### PR DESCRIPTION
**Pull request by lasers**
Add a formatter test for https://github.com/ultrabug/py3status/pull/1426. 

With `{'key': None}` and `format = '{key:.2f}'`... it prints `{key}`. (Before)
With `{'key': None}` and `format = '{key:.2f}'`... it prints `None`. (After)

It was faking `{nonexistent_key}` behavior instead of printing `None`.

### CODE REVIEW

**Comment by tobes** [comment]
The main issue I have here is this

```
>>>'{:.2f}'.format(None)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: Unknown format code 'f' for object of type 'str'
```

So it is about how much we try to support python formatting vs completely different.

For me I'd like to not fix these edge cases unless we are sure we want to keep supporting them.

I don't like breaking things for the user. 

How are you hitting this issue?  if it is a modules internals that is probably where to fix it

**Comment by ultrabug** [comment]
I sense a lot of trouble and discussions around the formatter guys.

Most of the disagreements come from the fact that @lasers wants to use it everywhere which leads to edge cases and @tobes wants to keep it simple and "in line" with Python's formatting.

The formatter is powerful but our frenzy around it slows us down. Am I the only one feeling this and am I wrong?

**Comment by tobes** [comment]
> our frenzy around it slows us down.

I'm with you on this - I'm also against this changing things for a single module without thinking things through - and yes I'm guilty of this too

I'm keen on trying to close old issues and re-review the PRs

**Comment by lasers** [comment]
Sorry, I didn't know it was an edge case. To me, it looks in line with other `test_numeric_*` functions in `test_formatter` (right above).

>How are you hitting this issue? if it is a modules internals that is probably where to fix it

I hit this in `lm_sensors` by doing `{None:g}`.
```python
        placeholders = self.py3.get_placeholders_list(self.format_sensor)
        self.format_sensor = self.py3.update_placeholder_formats(
            self.format_sensor, {x: ':g' for x in placeholders if x != 'name'}
        )
```

It's not possible to set a sane `default` format in `lm_sensors` so I'm making default format better for users by not trying to hide nonexistent `{input}` or not printing `{input}` itself on invalid sensors. I gave them `None` values so it'll show up `None` (more readable than several `{input}` if we were to use any other placeholders)... only to find that it does not change anything...  so that's how I found this issue.

**Comment by tobes** [comment]
> I didn't know it was an edge case. 

It might be better to say it is undefined behaviour and the nice thing about undefined behaviour is that you don't need to support it and can change things - the problem with undefined behaviour is that your users start using it and then you end up having to support it anyway :(

> I hit this in lm_sensors by doing {None:g}.

but why are you doing this?  To me if you are just changing all the placeholders to add formatting then my first question is why are we doing this?  If the problem is with the lm sensor output then this is what we should be looking at first to see if there is a nice solution.  this looks like you are brute forcing the issue.

So again if I understood the issue it'd help.  The fact that bad input gave an error is not the real problem I think

**Comment by lasers** [comment]
If you looked at `sensors -u`, the values are always `50.000`. I'm using `:g` to remove that in all placeholders except `{name}`. The placeholders are not always known and can vary from machine to machine. I harvest what I think is all. 

However, sensors vary from machine to machine too. Fan sensors != Cpu sensors != Voltage sensors != Unknown sensors != etc. Some placeholders does not belong in some sensors too.

That's when I fill in `None` instead of printing `{name}` because the only time users will see this is when they try it for first time and I'm keen on making default format really nice... so users don't always see `{nonexistent}` repeatedly in the long run because they don't have anything in their `lm_sensors`.

I do this for nice default but verbose `lm_sensors` output. I understand that the ideal way would be to leave them alone... To print `{name}` instead... but like I said, it's not possible to set a sane default `format`... so I'm keen on making default format nice for everybody with all different kind of hardware. 

Some users can have same hardware for 10 years so it's nice for that user not to see `{input}` everytime he load up `lm_sensors`. There are nothing wrong with their hardware since sensors often are different on every machine.

**Comment by lasers** [comment]
Try `sysdata` without `None`. That's what some users can see in default `lm_sensors` format.

```diff
diff --git a/py3status/modules/sysdata.py b/py3status/modules/sysdata.py
index eaea0afe..2399eed0 100644
--- a/py3status/modules/sysdata.py
+++ b/py3status/modules/sysdata.py
@@ -68,7 +68,11 @@ class Py3status:
     # available configuration parameters
     cache_timeout = 10
     format = "[\?color=cpu CPU: {cpu_used_percent}%], " \
-             "[\?color=mem Mem: {mem_used}/{mem_total} {mem_unit} ({mem_used_percent}%)]"
+             "[\?color=mem Mem: {mem_used}/{mem_total} {mem_unit} ({mem_used_percent}%)] " \
+             "[\?color=bad&show Fan:] {fan_input} " \
+             "[\?color=orange&show Voltage:] {input} " \
+             "[\?color=lightskyblue&show GPU: 40°C]"
+
     mem_unit = 'GiB'
     swap_unit = 'GiB'
     temp_unit = u'°C'

```

**Comment by tobes** [comment]
ok so you have at least 2 issues

1) you want to change the place holder formatting but don't know what placeholders will have numeric values

2) you want to hide missing placeholders?

so my thoughts are

1) don't change the format placeholders `:g` thing until you know they are numeric (just do on the first run when you have data)

2) missing ones you could either set them as `'None'` in the data sent to the formatter and not change the format for them.  leave as is which I would prefer as it is what all other modules do, or provide data like `this is not available :(`

so in summary I see nothing here requiring any change to the formatter.  The issue is just that the module does not care enough - ps is that really the output you get from `sensors -u` what is your version?

**Comment by lasers** [comment]
I think you don't have latest `lm_sensors` with JSON support so I tell you `sensors -u` instead... You can checkout the first `lm_sensors` commit to see `lm_sensors` live on your machine.

>don't change the format placeholders :g thing until you know they are numeric (just do on the first run when you have data)

I know they're all numeric except for `name`. I feed `None` to (missing) placeholders that does not have values because it's likely it had different placeholder names. Users are meant to filter them out eventually with `chips` and/or `sensors`.

**Comment by lasers** [comment]
>ps is that really the output you get from sensors -u what is your version?

Yes. First commit... I carve all lines up into a dictionary.... :-) 
Second commit... I use JSON output. No more carving the lines up.

**Comment by tobes** [comment]
```
>>> '{:.2f}'.format(7)
'7.00'
```
this is valid formatting `None` as a float isn't.  the point being that it is not valid in python and I think staying close to the python norms is good as that is the language we are using

now you could argue that allowing this to be ok is also bad, but I think here it is a reasonable conversion.
```
>>> '{:d}'.format(7.8)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: Unknown format code 'd' for object of type 'float'
```

**Comment by lasers** [comment]
To get what you want... I could work around them.
* Change `None` to `'None'` for `lm_sensors`.
* Remove `[ 'None']` for `nvidia_smi` in favor of checking each value versus checking output once.
* I can hardcode `.1f` for `memory.used_percent` so the percent can't be changed to `.2f` or `.0f`.
* Whatever else you don't like, mention them here.

**Comment by tobes** [comment]
Let's decide what we want to do here not just create a test

I am currently against this but maybe I can be persuaded by a good case - but let's make an issue and close this PR

**Comment by lasers** [comment]
No need for new issue. Keep this open instead.

`{None:.2f}` is already in `master`. Formatting errors is silent too in favor of values.

I don't have a strong feeling about this PR and am only adding a test to keep the tests updated with the formatter. This also make it clear what the test is for. I guess we could not merge this too.
